crates/ruff_linter/src/rules/ruff/rules/sort_dunder_all.rs

```diff
@@ -0,0 +1,882 @@
+        return;
+    }
+
+    let mut diagnostic = Diagnostic::new(UnsortedDunderAll, range);
+
+    if possibly_fixable {
+        if let Some(fix) = get_fix(range, elts, &string_items, &kind, checker) {
+            diagnostic.set_fix(fix);
+        }
+    }
+
+    checker.diagnostics.push(diagnostic);
+}
+
+enum DunderAllKind<'a> {
+    List,
+    Tuple(&'a ast::ExprTuple),
+}
+
+impl DunderAllKind<'_> {
+    fn opening_token_for_multiline_definition(&self) -> Tok {
+        match self {
+            Self::List => Tok::Lsqb,
+            Self::Tuple(_) => Tok::Lpar,
+        }
+    }
+
+    fn closing_token_for_multiline_definition(&self) -> Tok {
+        match self {
+            Self::List => Tok::Rsqb,
+            Self::Tuple(_) => Tok::Rpar,
+        }
+    }
+}
+
+fn dunder_all_is_already_sorted(string_elements: &[&str]) -> bool {
+    let mut element_iter = string_elements.iter();
+    let Some(this) = element_iter.next() else {
+        return true;
+    };
+    let mut this_key = AllItemSortKey::from(*this);
+    for next in element_iter {
+        let next_key = AllItemSortKey::from(*next);
+        if next_key < this_key {
+            return false;
+        }
+        this_key = next_key;
+    }
+    true
+}
+
+struct AllItemSortKey {
+    category: InferredMemberType,
+    value: String,
+}
+
+impl Ord for AllItemSortKey {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.category
+            .cmp(&other.category)
+            .then_with(|| natord::compare(&self.value, &other.value))
+    }
+}
+
+impl PartialOrd for AllItemSortKey {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for AllItemSortKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl Eq for AllItemSortKey {}
+
+impl From<&str> for AllItemSortKey {
+    fn from(value: &str) -> Self {
+        Self {
+            category: InferredMemberType::of(value),
+            value: String::from(value),
+        }
+    }
+}
+
+impl From<&DunderAllItem> for AllItemSortKey {
+    fn from(item: &DunderAllItem) -> Self {
+        Self::from(item.value.as_str())
+    }
+}
+
+/// Classification for an element in `__all__`.
+///
+/// This is necessary to achieve an "isort-style" sort,
+/// where elements are sorted first by category,
+/// then, within categories, are sorted according
+/// to a natural sort.
+///
+/// You'll notice that a very similar enum exists
+/// in ruff's reimplementation of isort.
+#[derive(Debug, Ord, PartialOrd, Eq, PartialEq, Clone, Copy)]
+enum InferredMemberType {
+    Constant,
+    Class,
+    Other,
+}
+
+impl InferredMemberType {
+    fn of(value: &str) -> Self {
+        // E.g. `CONSTANT`
+        if value.len() > 1 && is_cased_uppercase(value) {
+            Self::Constant
+        // E.g. `Class`
+        } else if value.chars().next().is_some_and(char::is_uppercase) {
```

Comment:
Nit
```suggestion
        } else if value.starts_with(char::is_uppercase) {
```

Reply:
Ah, this logic was purely copied-and-pasted from the `isort` rule 😄